datafusion/expr-common/src/signature.rs

```diff
@@ -996,13 +1159,119 @@ impl Signature {
                 },
             ),
             volatility,
+            parameter_names: None,
         }
     }
 
     /// Specialized [Signature] for ArrayEmpty and similar functions.
     pub fn array(volatility: Volatility) -> Self {
         Signature::arrays(1, Some(ListCoercion::FixedSizedListToList), volatility)
     }
+
+    /// Add parameter names to this signature, enabling named argument notation.
+    ///
+    /// # Example
+    /// ```
+    /// # use datafusion_expr_common::signature::{Signature, Volatility};
+    /// # use arrow::datatypes::DataType;
+    /// let sig = Signature::exact(vec![DataType::Int32, DataType::Utf8], Volatility::Immutable)
+    ///     .with_parameter_names(vec!["count".to_string(), "name".to_string()]);
+    /// ```
+    ///
+    /// # Errors
+    /// Returns an error if the number of parameter names doesn't match the signature's arity.
+    /// For signatures with variable arity (e.g., `Variadic`, `VariadicAny`), parameter names
+    /// cannot be specified.
+    pub fn with_parameter_names(
+        mut self,
+        names: Vec<String>,
+    ) -> datafusion_common::Result<Self> {
+        // Validate that the number of names matches the signature
+        self.validate_parameter_names(&names)?;
+        self.parameter_names = Some(names);
+        Ok(self)
+    }
+
+    /// Validate that parameter names are compatible with this signature
+    fn validate_parameter_names(
+        &self,
+        names: &[String],
+    ) -> datafusion_common::Result<()> {
+        // Get expected argument count from the type signature
+        let expected_count = match &self.type_signature {
```

Comment:
Maybe we should pull this expected count logic out of this function and put it as a function on `TypeSignature`:

```rust
let expected_count = self.type_signature.arity()
```

Though would need handling for variadic (maybe return an enum?)